app/main.py

```diff
@@ -1,34 +1,103 @@
+            is_drowned: bool = False
+    ) -> None:
+        self.decks = []
+        self.is_drowned = is_drowned
 
-    def get_deck(self, row, column):
-        # Find the corresponding deck in the list
-        pass
+        if start[0] == end[0]:
+            for column in range(start[1], end[1] + 1):
+                self.decks.append(Deck(start[0], column))
+        elif start[1] == end[1]:
+            for row in range(start[0], end[0] + 1):
+                self.decks.append(Deck(row, start[1]))
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def get_deck(self, row: int, column: int) -> Deck | None:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
+        return None
+
+    def fire(self, row: int, column: int) -> str:
+        deck = self.get_deck(row, column)
+        if deck:
+            deck.is_alive = False
+            if all(not _.is_alive for _ in self.decks):
+                self.is_drowned = True
+                return "Sunk!"
+            else:
+                return "Hit!"
+        return "Miss!"
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(
+            self,
+            ships: list[tuple[tuple[int, int], tuple[int, int]]]
+    ) -> None:
+        self.field = {}
+        self.ships = []
+
+        for start, end in ships:
+            ship = Ship(start, end)
+            self.ships.append(ship)
+            for deck in ship.decks:
+                self.field[(deck.row, deck.column)] = ship
+
+        self._validate_field()
+
+    def fire(self, location: tuple[int, int]) -> str:
+        if location in self.field:
+            ship = self.field[location]
+            result = ship.fire(location[0], location[1])
+            return result
+        return "Miss!"
+
+    def print_field(self) -> None:
+        field = [["~"] * 10 for _ in range(10)]
+
+        for ship in self.ships:
+            for deck in ship.decks:
+                if deck.is_alive:
+                    field[deck.row][deck.column] = "□"
+                else:
+                    field[deck.row][deck.column] = "*" if not ship.is_drowned\
+                        else "x"
+
+        for row in field:
+            print(" ".join(row))
+
+    def _validate_field(self) -> None:
+        if len(self.ships) != 10:
+            raise ValueError("Should be 10 ships.")
```

Comment:
The `_validate_field` method currently checks for exactly 10 ships, but it should also ensure that the distribution of ship sizes matches the required configuration: 1 ship of size 4, 2 ships of size 3, 3 ships of size 2, and 4 ships of size 1. This is critical to meet the task requirements.